File: src/main/java/org/example/observer/stockmarket2/StockMarket.java
package org.example.observer.stockmarket2;

import javax.swing.plaf.ListUI;
import java.util.ArrayList;
import java.util.List;

public class StockMarket implements Subject{

    private List<Observer> list;
    private String stockSymbol;
    private double stockPrice;
    public StockMarket(){
        list = new ArrayList<>();
    }

    public void setStockPriceStockSymbol(double stockPrice,String stockSymbol) {
        this.stockPrice = stockPrice;
        this.stockSymbol = stockSymbol;
        notifyObservers();
    }

    @Override
    public void registerObserver(Observer observer) {
        list.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        list.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : list) {
            observer.update(stockSymbol,stockPrice);
        }
    }
}

File: kotlin/src/main/kotlin/observer/StockMarket.kt
package observer

class StockMarket : Subject {
    val list = mutableListOf<Observer>()
    lateinit var stockSymbol: String
    var stockPrice: Int = 0

    fun setData(stockSymbol: String, stockPrice: Int) {
        this.stockSymbol = stockSymbol
        this.stockPrice = stockPrice
        notify1()
    }

    override fun registerObserver(observer: Observer) {
        list.add(observer)
    }

    override fun removeObserver(observer: Observer) {
        list.remove(observer)
    }

    override fun notify1() {
        for (observer in list) {
            observer.update(stockPrice,stockSymbol)
        }
    }
}